music/src/core/visualizer.ts

```diff
@@ -0,0 +1,199 @@
+   * automatically advancing the visualization if the note was painted outside
+   * of the screen.
+   */
+  redraw(activeNote?: NoteSequence.INote): number {
+    // TODO: this is not super optimal, and might start being too slow for
+    // larger sequences. Instead, we should figure out a way to store the
+    // "last painted active notes" and repaint those, as well as the new
+    // active notes instead.
+    this.ctx.clearRect(0, 0, this.ctx.canvas.width, this.ctx.canvas.height);
+    let activeNotePosition;
+
+    for (let i = 0; i < this.noteSequence.notes.length; i++) {
+      const note = this.noteSequence.notes[i];
+
+      // Size of this note.
+      const offset = this.config.noteSpacing * (i + 1);
+      const x = (this.getNoteStartTime(note) * this.config.pixelsPerTimeStep) +
+          offset;
+      const w = (this.getNoteEndTime(note) - this.getNoteStartTime(note)) *
+          this.config.pixelsPerTimeStep;
+
+      // The canvas' y=0 is at the top, but a smaller pitch is actually
+      // lower, so we're kind of painting backwards.
+      const y = this.height -
+          ((note.pitch - this.config.minPitch) * this.config.noteHeight);
+
+      // Color of this note.
+      const opacityBaseline = 0.2;  // Shift all the opacities up a little.
+      const opacity = note.velocity ? note.velocity / 100 + opacityBaseline : 1;
+
+      const isActive =
+          activeNote && this.isPaintingActiveNote(note, activeNote);
+      this.ctx.fillStyle =
+          `rgba(${isActive ? this.config.activeNoteRGB : this.config.noteRGB},
+          ${opacity})`;
+      this.ctx.fillRect(x, y, w, this.config.noteHeight);
+      if (isActive) {
+        activeNotePosition = x;
+      }
+    }
+    return activeNotePosition;
+  }
+
+  private getCanvasSize(): {width: number; height: number} {
+    // If the pitches haven't been specified already, figure them out
+    // from the NoteSequence.
+    if (this.config.minPitch === undefined ||
+        this.config.maxPitch === undefined) {
+      this.config.minPitch = 100;
+      this.config.maxPitch = -1;
+
+      // Find the smallest pitch so that we cans scale the drawing correctly.
+      for (const note of this.noteSequence.notes) {
+        if (note.pitch < this.config.minPitch) {
+          this.config.minPitch = note.pitch;
+        }
+        if (note.pitch > this.config.maxPitch) {
+          this.config.maxPitch = note.pitch;
+        }
+      }
+
+      // Add a little bit of padding at the top and the bottom;
+      this.config.minPitch -= 2;
+      this.config.maxPitch += 2;
+    }
+
+    // Height of the canvas based on the range of pitches in the sequence
+    const height =
+        (this.config.maxPitch - this.config.minPitch) * this.config.noteHeight;
+
+    // Calculate a nice width based on the length of the sequence we're playing.
+    const numNotes = this.noteSequence.notes.length;
+    const endTime = this.getNoteEndTime(this.noteSequence.notes[numNotes - 1]);
```

Comment:
We can't guarantee the notes are ordered, but there are a `total_time` and `total_quantized_steps` fields in the NoteSequence.